### DISCUSSION TITLE
Incorrect Typescript type Definitions

### DISCUSSION THREAD

**Discussion by gmixa** [nhn/tui.chart#609] [Q&A] [answered]
**Describe the bug**
I tried to use this project with scala-js and used the Typescript to Scala-js converter to creat Scala bindings.
Unfortunately that process Fails.

see:
https://github.com/ScalablyTyped/Converter/issues/249

The Reply that i got over there ist that somehow the included Typescript type definitions in this project are not Standards conform. As I am no Typescript expert i cannot give more Information on this aspect. But it seems as the parameters to create the charts cannot be set.
Maybe you can fix that issue in one of the next releases.
Greetings 
Gerald Mixa

Expected: A usable Type Definition in ScalaJS
Setup Description: see referenced bug

**Comment by jung-han** [comment]
Thank you for reporting the issue. After looking into it in detail, I will leave a reply again. Wait a bit!

**Comment by jung-han** [comment]
@gmixa 

Is it not that the typescript environment is a problem, but rather a problem when using the Scala js?
It would be nice if you let me know exactly what kind of problem there is. It seems that it is a little too difficult to identify the problem while using scala.

**Comment by gmixa** [comment]
Hi,
I have linked the Bug at the typescript to scala-js converter.
The Problem for example is that a method signature like this gets generated
def bulletChart(
props: /* import warning: transforms.QualifyReferences#resolveTypeRef many Couldn't qualify BulletChartProps */ js.Any
): BulletChart = js.native

The warning says that the parameters for creating a bulletChart could not be generated in a  correct manner.
@oyvindberg wrote in his comment about templating beeing invalid and did link to line 14 in tsconfig.json of this project. You defined a regex to match all files in your project there. I think that is causing the problems. He supposes that your typescript definitions also might be an issue for some typescript development setups as this should be illegal. I am no javascript/typescript expert and I hope I gave a good description to my best knowledge. To me it sounds reasonable that that could be an issue as i have not seen such a definition in other projects which i could use easily. Maybe you could ask @oyvindberg for a much better more detailed description as i think he has a deep insight into this topic as he maintains the type parser and evaluation project.

**Comment by oyvindberg** [comment]
The module names are... string templated. That will break ~every piece of tooling out there, including Typescript itself.

```
$ yarn add typescript @toast-ui/chart
$ cat test.ts
import foo from '@toast-ui/chart'

foo.columnChart(1)

$ node_modules/.bin/tsc test.ts 
node_modules/@toast-ui/chart/types/index.d.ts:54:8 - error TS2307: Cannot find module '@t/options' or its corresponding type declarations.

54 } from '@t/options';
          ~~~~~~~~~~~~
```

**Comment by jung-han** [comment]
@oyvindberg 
Sorry for the late reply. Would it be okay to understand that the `alias` we are using is a problem?
In this release, I will deploy by removing all aliases of types.

**Comment by oyvindberg** [comment]
Yes :)

**Comment by jung-han** [comment]
@oyvindberg @gmixa 
Resolved in [4.1.0](https://github.com/nhn/tui.chart/releases/tag/v4.1.0)! Check this version!

**Comment by gmixa** [comment]
I just tried @oyvindberg example.
It now causes the following issue:
npx tsc test.ts 
node_modules/@toast-ui/chart/types/components/axis.d.ts:9:8 - error TS2307: Cannot find module '../../src/brushes/label' or its corresponding type declarations.

9 } from '../../src/brushes/label';
         ~~~~~~~~~~~~~~~~~~~~~~~~~

node_modules/@toast-ui/chart/types/components/series.d.ts:2:48 - error TS2307: Cannot find module '../../src/brushes/basic' or its corresponding type declarations.

2 import { CircleStyleName, RectStyleName } from '../../src/brushes/basic';
                                                 ~~~~~~~~~~~~~~~~~~~~~~~~~

etc.

I think there is still something to do!

**Comment by jung-han** [comment]
@gmixa I'll check and fix it!

**Comment by jung-han** [comment]
@gmixa Hi! Are there any errors related to `src/component` or `src/store` other than brush related type errors? If possible, can you add more of the part where the error occurs?

**Comment by ilgwonPark2** [comment]
Hi @jung-han ,
I have a. same issue on ver 4.1.3.
It seems `src` folders are not included in the module.

Of course, including this folder is not compulsory, however, the module code is still referencing this path.
```
2:48 Cannot find module '../../src/brushes/basic' or its corresponding type declarations.
    1 | import { Point, Rect, BezierPoint, BoxSeriesDataType } from '../options';
  > 2 | import { CircleStyleName, RectStyleName } from '../../src/brushes/basic';
      |                                                ^
    3 | import { TooltipData } from './tooltip';
    4 | import { LineModel, LabelModel } from './axis';
    5 | import { SectorStyle, SectorStyleName } from '../../src/brushes/sector';

```

**Reply by gmixa** [reply]
My development Environment also shows that the import statements are missing. If i look at my setup it would be sufficient if the "src" part would be gone, I think. I have also tried tui-editor and tui-calendar with my setup. These seem to work with my setup. Maybe this information helps!

**Reply by jung-han** [reply]
@gmixa @ilgwonPark2 
The work has been completed and will be released as soon as today's review is completed. sorry to bother you!
#612

**Comment by jung-han** [comment]
@ilgwonPark2 
I'll check it immediately . Thank you for the issue.

**Reply by ilgwonPark2** [reply]
Can I track this issue in this discussion section?

**Reply by jung-han** [reply]
All of the `question` tag issues have been changed to Discussion.. For now, I would appreciate it if you track it here.😭  I will move on later.

**Comment by ilgwonPark2** [comment]
Thx! @jung-han :)

**Comment by jung-han** [comment, marked as answer]
@gmixa @ilgwonPark2 
Fixed in 4.1.4! Try it and let me know if you have a problem. I will fix it right away.
* https://github.com/nhn/tui.chart/releases/tag/v4.1.4

**Reply by ilgwonPark2** [reply]
Appreciate for It!
I will try it soon.

**Comment by gmixa** [comment]
Hi,
i did set up a  basic Example. I could draw a basic Line Chart. That works, even so i have to tune my own build process but that has nothing to do with your library. I think you managed to fix it. I just did no full regression test so no idea if every part is fully working. But from my point of view my original issue is fixed now! Thanks 👍

**Reply by jung-han** [reply]
Nice!
If you have any part to discuss later, please report new board at discussion or issue.
